cpp/src/traversal/extract_bfs_paths_impl.cuh

```diff
@@ -0,0 +1,378 @@
+
+    auto tuple_begin = thrust::make_zip_iterator(
+      thrust::make_tuple(tmp_frontier.begin(), original_position.begin(), original_rank.begin()));
+
+    std::forward_as_tuple(std::tie(tmp_frontier, original_position, original_rank), std::ignore) =
+      groupby_gpuid_and_shuffle_values(
+        handle.get_comms(),
+        tuple_begin,
+        tuple_begin + tmp_frontier.size(),
+        [key_func =
+           detail::mg_partition_vertices<vertex_t>{
+             handle.get_comms().get_rank(),
+             d_vertex_partition_offsets.begin(),
+             d_vertex_partition_offsets.end()}] __device__(auto val) {
+          return key_func(thrust::get<0>(val));
+        },
+        handle.get_stream());
+
+    tuple_begin = thrust::make_zip_iterator(
+      thrust::make_tuple(tmp_frontier.begin(), original_position.begin(), original_rank.begin()));
+
+    thrust::transform(
+      handle.get_thrust_policy(),
+      tuple_begin,
+      tuple_begin + tmp_frontier.size(),
+      tuple_begin,
+      detail::mg_lookup_distance<vertex_t>{graph_view.get_local_vertex_first(), distances});
+
+    std::forward_as_tuple(std::tie(tmp_frontier, original_position, original_rank), std::ignore) =
+      groupby_gpuid_and_shuffle_values(
+        handle.get_comms(),
+        tuple_begin,
+        tuple_begin + tmp_frontier.size(),
+        [] __device__(auto val) { return thrust::get<2>(val); },
+        handle.get_stream());
+
+    thrust::for_each_n(
+      handle.get_thrust_policy(),
+      thrust::make_zip_iterator(
+        thrust::make_tuple(tmp_frontier.begin(), original_position.begin())),
+      tmp_frontier.size(),
+      [max_path_length, d_current_position = current_position.data()] __device__(auto tuple) {
+        auto distance = thrust::get<0>(tuple);
+        auto position = thrust::get<1>(tuple);
+
+        size_t tmp                   = position;
+        d_current_position[position] = (tmp * max_path_length) + distance;
+      });
+  } else {
+    thrust::transform(
+      handle.get_thrust_policy(),
+      thrust::make_counting_iterator<size_t>(0),
+      thrust::make_counting_iterator<size_t>(n_destinations),
+      current_position.data(),
+      detail::map_index_to_path_offset<vertex_t>{max_path_length, distances, destinations});
+  }
+
+  std::tie(current_frontier, current_position) = detail::shrink_extraction_list(
+    handle, std::move(current_frontier), std::move(current_position));
+
+  thrust::for_each_n(handle.get_thrust_policy(),
+                     thrust::make_zip_iterator(
+                       thrust::make_tuple(current_frontier.begin(), current_position.begin())),
+                     current_frontier.size(),
+                     detail::update_paths<vertex_t>{paths, invalid_vertex});
+
+  size_t count =
+    multi_gpu
+      ? cugraph::host_scalar_allreduce(
+          handle.get_comms(), current_frontier.size(), raft::comms::op_t::MAX, handle.get_stream())
```

Comment:
Shouldn't this better be SUM reduction? (to minimize the difference in SG & MG logics)

Reply:
I think MAX is semantically what I want, although SUM would suffice.

I'm actually thinking I'll change this loop to a for loop from over the range [0, max_path_length), since we already know how many iterations are required.

Reply:
> I'm actually thinking I'll change this loop to a for loop from over the range [0, max_path_length), since we already know how many iterations are required.

+1 for this, this will be clearer.

Reply:
Changed to a for loop, definitely much clearer and saves the allreduce cost.